packages/core/crates/core-network/src/heartbeat.rs

```diff
@@ -1,80 +1,192 @@
-    use futures::stream::{Stream, StreamExt};
-    use js_sys::AsyncIterator;
-    use utils_misc::async_iterable::wasm::{to_box_u8_stream, to_jsvalue_stream};
-    use utils_misc::ok_or_jserr;
-    use utils_misc::utils::wasm::JsResult;
-    use utils_types::traits::BinarySerializable;
-    use wasm_bindgen::prelude::*;
-    use wasm_bindgen_futures::stream::JsStream;
-
-    #[wasm_bindgen]
-    impl HeartbeatConfig {
-        #[wasm_bindgen]
-        pub fn build(
-            max_parallel_heartbeats: usize,
-            heartbeat_variance: f32,
-            heartbeat_interval: u32,
-            heartbeat_threshold: u64,
-        ) -> Self {
-            Self {
-                max_parallel_heartbeats,
-                heartbeat_variance,
-                heartbeat_interval,
-                heartbeat_threshold,
-            }
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+impl HeartbeatConfig {
+    #[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen(constructor))]
+    pub fn new(heartbeat_variance: f32, heartbeat_interval: u32, heartbeat_threshold: u64) -> HeartbeatConfig {
+        HeartbeatConfig { heartbeat_variance, heartbeat_interval, heartbeat_threshold}
+    }
+}
+
+#[cfg_attr(test, mockall::automock)]
+#[async_trait]
+pub trait HeartbeatExternalApi {
+    async fn get_peers(&self, from_timestamp: u64) -> Vec<PeerId>;
+}
+
+
+pub struct Heartbeat<T: Pinging, API: HeartbeatExternalApi> {
+    config: HeartbeatConfig,
+    pinger: T,
+    external_api: API,
+    rng: rand::rngs::ThreadRng,
+    metric_time_to_heartbeat: Option<SimpleHistogram>,
+}
+
+impl<T: Pinging, API: HeartbeatExternalApi> Heartbeat<T, API> {
+    pub fn new(config: HeartbeatConfig, pinger: T, external_api: API) -> Self {
+        Self {
+            config, pinger, external_api,
+            rng: rand::thread_rng(),
+            metric_time_to_heartbeat: if cfg!(test) {None} else {SimpleHistogram::new(
+                "core_histogram_heartbeat_time_seconds",
+                "Measures total time it takes to probe all other nodes (in seconds)",
+                vec![0.5, 1.0, 2.5, 5.0, 10.0, 15.0, 30.0, 60.0, 90.0, 120.0, 300.0],
+            )
+            .ok()},
+        }
+    }
+
+    /// Heartbeat loop responsible for periodically requesting peers to ping around from the 
+    /// external API interface.
+    /// 
+    /// The loop never ends and will run indefinitely, until the program is explicitly terminated.
+    /// As such, this feature should therefore be joined with other internal loops and awaited
+    /// after all components have been initialized.
+    pub async fn heartbeat_loop(&mut self) {
+        loop {
+            let heartbeat_round_timer = if let Some(metric_time_to_heartbeat) = &self.metric_time_to_heartbeat {
+                Some(histogram_start_measure!(metric_time_to_heartbeat))
+            } else {
+                None
+            };
```

Comment:
This is essentially `Option::map`

Reply:
There could be a more generalized approach using the `scope_guard`, not an option.